numpy/lib/function_base.py

```diff
@@ -3912,37 +3991,86 @@ def quantile(a, q, axis=None, out=None,
     >>> assert not np.all(a == b)
     """
     q = np.asanyarray(q)
+    
     if not _quantile_is_valid(q):
         raise ValueError("Quantiles must be in the range [0, 1]")
+        
+    if w is not None:
+        a = np.asanyarray(a)
+        w = np.asanyarray(w)
+
+        if w.shape != a.shape:
+            if w.ndim != 1:
+                raise TypeError(
+                    "1D weights expected when shapes of a and weights differ."
+                )
+            if axis is None:
+                if w.size != a.size:
+                    raise TypeError(
+                        "Length of weights not compatible with a's size"
+                    )
+            else:
+                if w.shape[0] != a.shape[axis]:
+                    raise ValueError(
+                        "Length of weights not compatible with specified axis."
+                    )
+                w = np.broadcast_to(w, (a.ndim-1) * (1,) + w.shape)
+                w = w.swapaxes(-1, axis) * np.ones(a.shape)
+        if not _weight_is_valid(w):
+            raise ValueError("All the weights must be positive")
+
     return _quantile_unchecked(
-        a, q, axis, out, overwrite_input, interpolation, keepdims)
+        a, q, w, axis, out, overwrite_input, interpolation, keepdims)
 
 
-def _quantile_unchecked(a, q, axis=None, out=None, overwrite_input=False,
-                        interpolation='linear', keepdims=False):
+def _quantile_unchecked(a, q, w=None, axis=None, out=None,
+                        overwrite_input=False, interpolation='linear',
+                        keepdims=False):
     """Assumes that q is in [0, 1], and is an ndarray"""
-    r, k = _ureduce(a, func=_quantile_ureduce_func, q=q, axis=axis, out=out,
-                    overwrite_input=overwrite_input,
-                    interpolation=interpolation)
+    if w is None:
+        r, k = _ureduce(a, func=_quantile_ureduce_func, q=q, axis=axis, out=out,
+                        overwrite_input=overwrite_input,
+                        interpolation=interpolation)
+    else:
+        r, k = _weighted_ureduce(a, func=_weighted_quantile_ureduce_func, w=w,
+                                 q=q, axis=axis, out=out,
+                                 overwrite_input=overwrite_input,
+                                 interpolation=interpolation)
     if keepdims:
         return r.reshape(q.shape + k)
     else:
         return r
 
 
 def _quantile_is_valid(q):
-    # avoid expensive reductions, relevant for arrays with < O(1000) elements
+    # Avoid expensive reductions, relevant for arrays with < O(1000) elements
```

Comment:
Irrelevant code should not be changed.

Reply:
Following PEP 8,  the first word in comments should be capitalized.
I confused whether to modify it.

Reply:
It's better to create a new PR to revise it. Not in this PR.

Reply:
It's fine to do this here and there in this PR, as long as there are not too many changes like this.